File: packages/sdk/dart/packages/flutter/lib/src/token_manager.dart
// Token management for EdgeBase SDK.
//
// Access + refresh tokens are persisted as one pair so a response-loss restart
// can replay an anonymous-account upgrade checkpoint safely.

import 'dart:async';
import 'dart:convert';
import 'package:edgebase_core/src/token_manager.dart' as core;
import 'package:edgebase_core/src/errors.dart';

import 'shared_prefs_token_storage_stub.dart'
    if (dart.library.ui) 'shared_prefs_token_storage_flutter.dart';
import 'token_storage.dart';

export 'token_storage.dart';

/// User info extracted from JWT.
class TokenUser {
  final String id;
  final String? email;
  final String? displayName;
  final String? avatarUrl;
  final bool? emailVerified;
  final bool? isAnonymous;
  final Map<String, dynamic>? customClaims;

  TokenUser({
    required this.id,
    this.email,
    this.displayName,
    this.avatarUrl,
    this.emailVerified,
    this.isAnonymous,
    this.customClaims,
  });

  factory TokenUser.fromJwtPayload(Map<String, dynamic> payload) {
    return TokenUser(
      id: payload['sub'] as String,
      email: payload['email'] as String?,
      displayName: payload['displayName'] as String?,
      avatarUrl: payload['avatarUrl'] as String?,
      emailVerified: payload['emailVerified'] as bool?,
      isAnonymous: payload['isAnonymous'] as bool?,
      customClaims: payload['customClaims'] as Map<String, dynamic>?,
    );
  }
}

/// Token pair returned from refresh.
class TokenPair {
  final String accessToken;
  final String refreshToken;
  TokenPair({required this.accessToken, required this.refreshToken});
}

/// Callback type for performing token refresh via HTTP.
typedef RefreshTokenCallback = Future<TokenPair> Function(String refreshToken);

/// Persistent token storage failed before a new session could be exposed.
class TokenPersistenceException implements Exception {
  final String operation;
  final Object cause;
  final StackTrace causeStackTrace;

  TokenPersistenceException(
    this.operation,
    this.cause,
    this.causeStackTrace,
  );

  @override
  String toString() =>
      'TokenPersistenceException: $operation failed before token adoption: $cause';
}

class InvalidTokenPairException implements Exception {
  final String operation;
  const InvalidTokenPairException(this.operation);

  @override
  String toString() =>
      'InvalidTokenPairException: incomplete token pair during $operation';
}

/// Token manager — handles Access/Refresh tokens and auth state.
/// Implements core.TokenManager so it can be used with HttpClient and other
/// core components that accept the abstract interface.
class TokenManager implements core.TokenManager {
  final String baseUrl;
  final TokenStorage storage;
  String? _accessToken;
  int? _accessTokenExp; // JWT exp claim (seconds since epoch)
  TokenUser? _currentUser;
  final _authStateController = StreamController<TokenUser?>.broadcast();
  bool _isRefreshing = false;
  bool _isClosed = false;  // guards against add-after-close
  Completer<String?>? _refreshCompleter;

  /// Buffer in seconds before actual expiry to trigger preemptive refresh.
  static const int _expiryBufferSeconds = 30;

  TokenManager({
    required this.baseUrl,
    TokenStorage? storage,
  }) : storage = storage ?? SharedPrefsTokenStorage();

  /// Current access token (in memory only).
  String? get accessToken => _accessToken;

  /// Current user parsed from JWT.
  TokenUser? get currentUser => _currentUser;

  /// Broadcast stream of auth state changes.
  /// Use [currentUser] for the initial state — the stream only emits on
  /// subsequent changes (matching the Firebase/Dart convention).
  Stream<TokenUser?> get onAuthStateChange => _authStateController.stream;

  /// Check if the current access token is expired (with 30s buffer).
  bool get isTokenExpired {
    if (_accessTokenExp == null) return true;
    final now = DateTime.now().millisecondsSinceEpoch ~/ 1000;
    return now >= (_accessTokenExp! - _expiryBufferSeconds);
  }

  /// Get the stored refresh token (implements core.TokenManager).
  @override
  Future<String?> getRefreshToken() async =>
      (await _loadStoredPair())?.refreshToken;

  /// Get a valid access token, refreshing if needed.
  /// Implements core.TokenManager with optional RefreshCallback.
  @override
  Future<String?> getAccessToken([core.RefreshCallback? refreshCallback]) async {
    // If we have a valid (non-expired) token, return it immediately
    if (_accessToken != null && !isTokenExpired) {
      return _accessToken;
    }

    if (refreshCallback == null) return _accessToken;

    // If already refreshing, wait for that to complete
    if (_isRefreshing && _refreshCompleter != null) {
      return _refreshCompleter!.future;
    }

    // Try to refresh
    final storedRefreshToken = (await _loadStoredPair())?.refreshToken;
    if (storedRefreshToken == null) return _accessToken;

    _isRefreshing = true;
    _refreshCompleter = Completer<String?>();

    try {
      final pair = await refreshCallback(storedRefreshToken);
      await _applyTokens(
        accessToken: pair.accessToken,
        refreshToken: pair.refreshToken,
      );
      _refreshCompleter!.complete(_accessToken);
      return _accessToken;
    } catch (e) {
      if (e is TokenPersistenceException || e is InvalidTokenPairException) {
        final completion = _refreshCompleter!;
        // Attach a handler for the initiating caller while still propagating
        // the same failure to any concurrent refresh waiters.
        unawaited(completion.future.catchError((_) => null));
        final failureStack = e is TokenPersistenceException
            ? e.causeStackTrace
            : StackTrace.current;
        completion.completeError(e, failureStack);
        rethrow;
      }
      // 401 means token revoked/expired — clear session (matches JS SDK).
      // Other errors (network, 5xx) keep session for retry.
      if (e is EdgeBaseError && e.statusCode == 401) {
        await clearTokens();
        _refreshCompleter!.complete(null);
        return null;
      }
      _refreshCompleter!.complete(_accessToken);
      return _accessToken;
    } finally {
      _isRefreshing = false;
      _refreshCompleter = null;
    }
  }

  /// Set tokens (core.TokenManager interface: positional args).
  @override
  Future<void> setTokens(String access, String refresh) =>
      _applyTokens(accessToken: access, refreshToken: refresh);

  /// Internal helper: applies token pair with named params.
  Future<void> _applyTokens({
    required String accessToken,
    required String refreshToken,
  }) async {
    if (accessToken.isEmpty || refreshToken.isEmpty) {
      throw const InvalidTokenPairException('save');
    }
    final nextUser = _decodeJwt(accessToken);
    final nextExpiration = _extractExp(accessToken);
    try {
      if (storage is AtomicTokenPairStorage) {
        await (storage as AtomicTokenPairStorage).setTokenPair(StoredTokenPair(
          accessToken: accessToken,
          refreshToken: refreshToken,
        ));
      } else {
        await storage.setRefreshToken(refreshToken);
      }
    } catch (error, stackTrace) {
      throw TokenPersistenceException('save', error, stackTrace);
    }
    // Persist-before-expose: a failed refresh-token write must never leave a
    // new access token or user visible only in memory.
    _accessToken = accessToken;
    _currentUser = nextUser;
    _accessTokenExp = nextExpiration;
    if (!_isClosed) _authStateController.add(_currentUser);
  }

  void _adoptStoredTokens(StoredTokenPair pair) {
    final accessToken = pair.accessToken;
    if (accessToken == null ||
        accessToken.isEmpty ||
        pair.refreshToken.isEmpty) {
      throw const InvalidTokenPairException('restore');
    }
    _accessToken = accessToken;
    _currentUser = _decodeJwt(accessToken);
    _accessTokenExp = _extractExp(accessToken);
    if (!_isClosed) _authStateController.add(_currentUser);
  }

  Future<StoredTokenPair?> _loadStoredPair() async {
    try {
      if (storage is AtomicTokenPairStorage) {
        return await (storage as AtomicTokenPairStorage).getTokenPair();
      }
      final refreshToken = await storage.getRefreshToken();
      return refreshToken == null
          ? null
          : StoredTokenPair(accessToken: null, refreshToken: refreshToken);
    } catch (error, stackTrace) {
      throw TokenPersistenceException('load', error, stackTrace);
    }
  }

  /// Fail closed before an operation that can revoke the initiating session.
  void requireDurableStorageForAccountUpgrade() {
    if (storage is DurableTokenStorage || currentUser?.isAnonymous == false) {
      return;
    }
    throw StateError(
      'Anonymous account upgrades require durable access/refresh pair storage '
      'so replacement tokens survive response loss or process termination.',
    );
  }

  /// Clear tokens on sign-out.
  Future<void> clearTokens() async {
    try {
      await storage.clearRefreshToken();
    } catch (error, stackTrace) {
      throw TokenPersistenceException('clear', error, stackTrace);
    }
    _accessToken = null;
    _accessTokenExp = null;
    _currentUser = null;
    if (!_isClosed) _authStateController.add(null);
  }

  /// Decode JWT payload (no verification — server is source of truth).
  TokenUser? _decodeJwt(String jwt) {
    try {
      final parts = jwt.split('.');
      if (parts.length != 3) return null;
      final payload = parts[1];
      final normalized = base64Url.normalize(payload);
      final decoded = utf8.decode(base64Url.decode(normalized));
      final json = jsonDecode(decoded) as Map<String, dynamic>;
      return TokenUser.fromJwtPayload(json);
    } catch (_) {
      return null;
    }
  }

  /// Extract exp claim from JWT.
  int? _extractExp(String jwt) {
    try {
      final parts = jwt.split('.');
      if (parts.length != 3) return null;
      final payload = parts[1];
      final normalized = base64Url.normalize(payload);
      final decoded = utf8.decode(base64Url.decode(normalized));
      final json = jsonDecode(decoded) as Map<String, dynamic>;
      return json['exp'] as int?;
    } catch (_) {
      return null;
    }
  }

  /// Try to restore session from stored refresh token.
  Future<bool> tryRestoreSession(RefreshTokenCallback refreshFn) async {
    final stored = await _loadStoredPair();
    if (stored == null) return false;

    if (stored.accessToken != null) {
      _adoptStoredTokens(stored);
      return true;
    }

    try {
      final pair = await refreshFn(stored.refreshToken);
      await _applyTokens(
        accessToken: pair.accessToken,
        refreshToken: pair.refreshToken,
      );
      return true;
    } catch (error) {
      if (error is TokenPersistenceException ||
          error is InvalidTokenPairException) {
        rethrow;
      }
      await clearTokens();
      return false;
    }
  }

  void destroy() {
    _isClosed = true;
    _authStateController.close();
  }
}

File: packages/sdk/swift/packages/ios/Sources/TokenManager.swift
import EdgeBaseCore
// Token Manager — JWT token management with auto-refresh.
//
// Mirrors Dart SDK TokenManager with Swift idioms:
// - Keychain persistence via protocol (DI)
// - 30-second buffer preemptive refresh
// - Actor for thread-safe concurrent refresh deduplication

import Foundation

// MARK: - Token Storage Protocol (DI —)

/// Protocol for persistent token storage.
/// Default implementation uses Keychain. Override for testing.
public protocol TokenStorage: Sendable {
    func getTokens() async throws -> TokenPair?
    func saveTokens(_ tokens: TokenPair) async throws
    func clearTokens() async throws
}

/// Storage that survives process restart and reports failed writes.
public protocol DurableTokenStorage: TokenStorage {}

/// Token pair (access + refresh).
public struct TokenPair: Codable, Sendable {
    public let accessToken: String
    public let refreshToken: String

    public init(accessToken: String, refreshToken: String) {
        self.accessToken = accessToken
        self.refreshToken = refreshToken
    }
}

/// Callback type for refreshing tokens.
public typealias RefreshTokenCallback = @Sendable (String) async throws -> TokenPair

// MARK: - In-Memory Token Storage (for testing)

/// Simple in-memory token storage for testing.
public actor MemoryTokenStorage: TokenStorage {
    private var tokens: TokenPair?

    public init() {}

    public func getTokens() async -> TokenPair? { tokens }
    public func saveTokens(_ tokens: TokenPair) async { self.tokens = tokens }
    public func clearTokens() async { tokens = nil }
}

// MARK: - Keychain Token Storage

/// Keychain-based persistent token storage.
public final class KeychainTokenStorage: DurableTokenStorage, @unchecked Sendable {
    private let service: String
    private let accessGroup: String?
    private let queue = DispatchQueue(label: "com.edgebase.keychain")

    public init(service: String = "com.edgebase.tokens", accessGroup: String? = nil) {
        self.service = service
        self.accessGroup = accessGroup
    }

    public func getTokens() async throws -> TokenPair? {
        try queue.sync {
            var query = baseQuery()
            query[kSecReturnData as String] = true
            query[kSecMatchLimit as String] = kSecMatchLimitOne

            var result: AnyObject?
            let status = SecItemCopyMatching(query as CFDictionary, &result)
            if status == errSecItemNotFound { return nil }
            guard status == errSecSuccess else {
                throw KeychainTokenStorageError(operation: "read", status: status)
            }
            guard let data = result as? Data else {
                throw KeychainTokenStorageError(operation: "read-invalid-data", status: status)
            }
            do {
                return try JSONDecoder().decode(TokenPair.self, from: data)
            } catch {
                throw KeychainTokenStorageError(operation: "decode", status: errSecDecode)
            }
        }
    }

    public func saveTokens(_ tokens: TokenPair) async throws {
        try queue.sync {
            let data: Data
            do {
                data = try JSONEncoder().encode(tokens)
            } catch {
                throw KeychainTokenStorageError(operation: "encode", status: errSecParam)
            }

            let query = baseQuery()

            // Try update first, then add
            let updateAttrs: [String: Any] = [kSecValueData as String: data]
            let updateStatus = SecItemUpdate(query as CFDictionary, updateAttrs as CFDictionary)

            if updateStatus == errSecItemNotFound {
                var addQuery = query
                addQuery[kSecValueData as String] = data
                addQuery[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
                let addStatus = SecItemAdd(addQuery as CFDictionary, nil)
                guard addStatus == errSecSuccess else {
                    throw KeychainTokenStorageError(operation: "add", status: addStatus)
                }
            } else if updateStatus != errSecSuccess {
                throw KeychainTokenStorageError(operation: "update", status: updateStatus)
            }
        }
    }

    public func clearTokens() async throws {
        try queue.sync {
            let query = baseQuery()
            let status = SecItemDelete(query as CFDictionary)
            guard status == errSecSuccess || status == errSecItemNotFound else {
                throw KeychainTokenStorageError(operation: "delete", status: status)
            }
        }
    }

    private func baseQuery() -> [String: Any] {
        var query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: "tokens",
        ]
        if let group = accessGroup {
            query[kSecAttrAccessGroup as String] = group
        }
        return query
    }
}

public struct KeychainTokenStorageError: Error, LocalizedError, Sendable, Equatable {
    public let operation: String
    public let status: OSStatus

    public var errorDescription: String? {
        "Apple Keychain \(operation) failed (OSStatus \(status))."
    }
}

public struct InvalidTokenPairError: Error, LocalizedError, Sendable, Equatable {
    public let operation: String
    public var errorDescription: String? {
        "EdgeBase token pair is incomplete during \(operation)."
    }
}

public struct DurableTokenStorageRequiredError: Error, LocalizedError, Sendable, Equatable {
    public init() {}
    public var errorDescription: String? {
        "Anonymous account upgrades require DurableTokenStorage so replacement tokens survive response loss or process termination."
    }
}

public struct TokenPersistenceError: Error, LocalizedError, Sendable, Equatable {
    public let operation: String
    public let causeDescription: String

    public init(operation: String, cause: Error) {
        self.operation = operation
        self.causeDescription = String(describing: cause)
    }

    public var errorDescription: String? {
        "Token persistence \(operation) failed before token adoption: \(causeDescription)"
    }
}

// MARK: - Token Manager

/// Manages access/refresh tokens with auto-refresh.
public actor TokenManager: TokenManageable {
    private let storage: TokenStorage
    private var currentTokens: TokenPair?
    private var refreshCallback: RefreshTokenCallback?
    private var refreshTask: Task<TokenPair, Error>?
    /// Bumped on every logout (clearTokens). An in-flight refresh compares the
    /// generation it started with against the current one; if they differ, a logout
    /// happened mid-refresh and the refreshed tokens must NOT resurrect the session.
    private var sessionGeneration = 0

    // Auth state change stream
    private var authStateHandlers: [(([String: Any]?) -> Void)] = []

    public init(storage: TokenStorage) {
        self.storage = storage
    }

    /// Set the refresh callback.
    public func setRefreshCallback(_ callback: @escaping RefreshTokenCallback) {
        self.refreshCallback = callback
    }

    /// Set tokens after login.
    public func setTokens(_ tokens: TokenPair) async throws {
        guard !tokens.accessToken.isEmpty, !tokens.refreshToken.isEmpty else {
            throw InvalidTokenPairError(operation: "save")
        }
        do {
            try await storage.saveTokens(tokens)
        } catch {
            throw TokenPersistenceError(operation: "save", cause: error)
        }
        currentTokens = tokens
        let user = decodeJWTPayload(tokens.accessToken)
        notifyAuthStateChange(user)
    }

    /// Clear tokens (logout).
    ///
    /// Cancels any in-flight refresh and bumps the session generation so a refresh
    /// that is already awaiting the network cannot re-persist tokens or emit a
    /// signed-in event after logout.
    public func clearTokens() async throws {
        sessionGeneration &+= 1
        refreshTask?.cancel()
        refreshTask = nil
        do {
            try await storage.clearTokens()
        } catch {
            throw TokenPersistenceError(operation: "clear", cause: error)
        }
        currentTokens = nil
        notifyAuthStateChange(nil)
    }

    /// Try to restore session from storage.
    /// Notifies auth state listeners on success (matches JS SDK auto-restore behavior).
    public func tryRestoreSession() async throws -> Bool {
        let stored: TokenPair?
        do {
            stored = try await storage.getTokens()
        } catch {
            throw TokenPersistenceError(operation: "load", cause: error)
        }
        if let tokens = stored {
            guard !tokens.accessToken.isEmpty, !tokens.refreshToken.isEmpty else {
                throw InvalidTokenPairError(operation: "restore")
            }
            currentTokens = tokens
            let user = decodeJWTPayload(tokens.accessToken)
            notifyAuthStateChange(user)
            return true
        }
        return false
    }

    /// Fail closed before an operation that can revoke the initiating session.
    public func requireDurableStorageForAccountUpgrade() throws {
        if storage is any DurableTokenStorage { return }
        if currentUser()?["isAnonymous"] as? Bool == false { return }
        throw DurableTokenStorageRequiredError()
    }

    /// Get valid access token, refreshing if needed.
    /// Includes 30-second buffer for preemptive refresh.

    /// Get current refresh token.
    public func getRefreshToken() -> String? {
        return currentTokens?.refreshToken
    }

    /// Get current user from cached access token (decoded JWT payload).
    public func currentUser() -> [String: Any]? {
        guard let token = currentTokens?.accessToken else { return nil }
        return decodeJWTPayload(token)
    }
    public func getAccessToken() async throws -> String? {
        return try await getAccessToken(forceRefresh: false)
    }

    /// Get valid access token, refreshing if needed.
    /// When `forceRefresh` is true, skip the local-expiry short-circuit and refresh
    /// unconditionally (used on a server 401 so the retry uses a freshly minted token
    /// instead of re-sending one the server already rejected).
    public func getAccessToken(forceRefresh: Bool) async throws -> String? {
        guard let tokens = currentTokens else { return nil }

        // Check if token is expired or will expire within 30s
        if !forceRefresh && !isTokenExpired(tokens.accessToken) {
            return tokens.accessToken
        }

        // Deduplicate concurrent refresh requests
        if let existingTask = refreshTask {
            let newTokens = try await existingTask.value
            return newTokens.accessToken
        }

        guard let refreshCb = refreshCallback else {
            return tokens.accessToken
        }

        let generation = sessionGeneration
        let refreshToken = tokens.refreshToken
        let task = Task<TokenPair, Error> { [weak self] in
            guard let self else {
                throw EdgeBaseError(statusCode: 401, message: "Token manager was released during refresh.")
            }
            let newTokens = try await refreshCb(refreshToken)
            // If a logout happened while we were awaiting the network, do NOT
            // resurrect the session by persisting tokens or firing signed-in events.
            let committed = try await self.commitRefreshedTokens(newTokens, generation: generation)
            guard committed else {
                throw EdgeBaseError(statusCode: 401, message: "Session was cleared during token refresh.")
            }
            return newTokens
        }
        refreshTask = task

        defer { refreshTask = nil }
        do {
            let newTokens = try await task.value
            return newTokens.accessToken
        } catch {
            try await handleRefreshFailure(error, generation: generation)
            throw error
        }
    }

    /// Persist refreshed tokens and emit a signed-in event, unless a logout has
    /// occurred since the refresh started. Returns false when the refresh is stale.
    private func commitRefreshedTokens(_ newTokens: TokenPair, generation: Int) async throws -> Bool {
        guard generation == sessionGeneration else { return false }
        guard !newTokens.accessToken.isEmpty, !newTokens.refreshToken.isEmpty else {
            throw InvalidTokenPairError(operation: "refresh")
        }
        do {
            try await storage.saveTokens(newTokens)
        } catch {
            throw TokenPersistenceError(operation: "save", cause: error)
        }
        guard generation == sessionGeneration else { return false }
        currentTokens = newTokens
        notifyAuthStateChange(decodeJWTPayload(newTokens.accessToken))
        return true
    }

    /// On a refresh failure with 401, clear the session (token revoked/expired),
    /// matching the JS reference. Other errors (network, 5xx) keep the session for
    /// retry. Skipped if a logout already advanced the generation.
    private func handleRefreshFailure(_ error: Error, generation: Int) async throws {
        guard (error as? EdgeBaseError)?.statusCode == 401 else { return }
        guard generation == sessionGeneration else { return }
        try await clearTokens()
    }

    /// Check if token is expired (with 30s buffer).
    public func isTokenExpired(_ token: String) -> Bool {
        guard let payload = decodeJWTPayload(token),
              let exp = payload["exp"] as? TimeInterval else {
            return true
        }
        // 30-second buffer for preemptive refresh
        return Date().timeIntervalSince1970 >= (exp - 30)
    }

    /// Register auth state change handler.
    /// If there is an active session, fires immediately with the current user
    /// (matches JS SDK behavior). Nil is not emitted on registration since it's
    /// the default state and would race with async Task-based callers like RoomClient.
    public func onAuthStateChange(_ handler: @escaping ([String: Any]?) -> Void) {
        authStateHandlers.append(handler)
        if let user = currentUser() {
            handler(user)
        }
    }

    /// Notify all auth state change handlers.
    private func notifyAuthStateChange(_ user: [String: Any]?) {
        for handler in authStateHandlers {
            handler(user)
        }
    }

    /// Decode JWT payload (base64url → JSON).
    private func decodeJWTPayload(_ token: String) -> [String: Any]? {
        let parts = token.split(separator: ".")
        guard parts.count >= 2 else { return nil }

        var base64 = String(parts[1])
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")

        // Pad to multiple of 4
        while base64.count % 4 != 0 { base64.append("=") }

        guard let data = Data(base64Encoded: base64),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }
        var normalized = json
        if normalized["id"] == nil {
            normalized["id"] = normalized["sub"] ?? normalized["userId"]
        }
        if normalized["customClaims"] == nil, let custom = normalized["custom"] as? [String: Any] {
            normalized["customClaims"] = custom
        }
        return normalized
    }

    /// Destroy — clean up handlers.
    public func destroy() {
        authStateHandlers.removeAll()
        refreshTask?.cancel()
        refreshTask = nil
    }
}

// MARK: - External Token Manager (closure-based, for roomWithToken)

/// A TokenManageable that delegates token retrieval to a closure.
/// Used by EdgeBaseClient.roomWithToken to accept an external token provider.
public actor ExternalTokenManager: TokenManageable {
    private let provider: @Sendable () -> String

    public init(tokenProvider: @escaping @Sendable () -> String) {
        self.provider = tokenProvider
    }

    public func getAccessToken() async throws -> String? {
        let token = provider()
        return token.isEmpty ? nil : token
    }

    public func getRefreshToken() async -> String? { nil }
    public func clearTokens() async {}
}
